src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,571 @@
+    options: S3Options,
+) -> Request {
+    let (inner, key) = req.into_parts();
+
+    let filename = Uuid::new_v4().to_string();
+
+    let key = String::from_utf8_lossy(&key[..]).into_owned();
+    let key = format!(
+        "{}/{}{}.{}",
+        path_prefix.unwrap_or_default(),
+        key,
+        filename,
+        "json.gz"
+    )
+    .replace("//", "/");
+
+    debug!(
+        message = "Sending events.",
+        bytes = ?inner.len(),
+        bucket = ?bucket,
+        key = ?key
+    );
+
+    Request {
+        body: inner,
+        bucket,
+        key,
+        content_encoding: Compression::gzip_default().content_encoding(),
+        options: aws_s3::S3Options {
+            acl: options.acl,
+            grant_full_control: options.grant_full_control,
+            grant_read: options.grant_read,
+            grant_read_acp: options.grant_read_acp,
+            grant_write_acp: options.grant_write_acp,
+            server_side_encryption: options.server_side_encryption,
+            ssekms_key_id: options.ssekms_key_id,
+            storage_class: options.storage_class,
+            tags: options.tags,
+            content_encoding: None,
+            content_type: None,
+        },
+    }
+}
+
+/// Applies the following transformations to align event's schema with DD:
+/// - `_id` is generated in the sink(format described below);
+/// - `date` is set from the Global Log Schema's `timestamp` mapping, or to the current time if missing;
+/// - `message`,`host` are set from the corresponding Global Log Schema mappings;
+/// - `source`, `service`, `status`, `tags` and other reserved attributes are left as is;
+/// - the rest of the fields is moved to `attributes`.
+fn encode_event(event: Event) -> Option<EncodedEvent<PartitionInnerBuffer<Vec<u8>, Bytes>>> {
+    lazy_static! {
+        static ref KEY_PREFIX: Template =
+            Template::try_from("/dt=%Y%m%d/hour=%H/").expect("invalid object key format");
+    }
+    let key = KEY_PREFIX
+        .render_string(&event)
+        .map_err(|error| {
+            emit!(TemplateRenderingFailed {
+                error,
+                field: Some("key_prefix"),
+                drop_event: true,
+            });
+        })
+        .ok()?;
+
+    let mut log = event.into_log();
+
+    log.insert("_id", generate_log_id());
+
+    let timestamp = log
+        .remove(crate::config::log_schema().timestamp_key())
+        .unwrap_or_else(|| chrono::Utc::now().timestamp_millis().into());
+    log.insert(
+        "date",
+        timestamp
+            .as_timestamp()
+            .cloned()
+            .unwrap_or_else(chrono::Utc::now)
+            .to_rfc3339_opts(SecondsFormat::Millis, true),
+    );
+    if let Some(message) = log.remove(crate::config::log_schema().message_key()) {
+        log.insert("message", message);
+    }
+    if let Some(message) = log.remove(crate::config::log_schema().host_key()) {
+        log.insert("host", message);
+    }
```

Comment:
Because `log_schema()` goes through a `lazy_static` it incurs an atomic barrier, no good for a hot function. Please put `encode_event` into a struct and stash the keys you need in that struct. You can see where I've done something similar [here](https://github.com/timberio/vector/pull/8789/files).

Reply:
Also consider the use of `rename_key` introduced in #8907.

Reply:
Done, also used `rename_key_flat` 👍